src/modules/ekf2/EKF/fake_pos_control.cpp

```diff
@@ -0,0 +1,123 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file fake_pos_control.cpp
+ * Control functions for ekf fake position fusion
+ */
+
+#include "ekf.h"
+
+void Ekf::controlFakePosFusion()
+{
+	// If we aren't doing any aiding, fake position measurements at the last known position to constrain drift
+	// During intial tilt aligment, fake position is used to perform a "quasi-stationary" leveling of the EKF
+	const bool fake_pos_data_ready = isTimedOut(_time_last_hor_pos_fuse, (uint64_t)2e5); // Fuse fake position at a limited rate
+
+	if (fake_pos_data_ready) {
+		const bool continuing_conditions_passing = !isHorizontalAidingActive() && !(_control_status.flags.fuse_aspd && _control_status.flags.fuse_beta);
+		const bool starting_conditions_passing = continuing_conditions_passing;
+
+		if (_using_synthetic_position) {
+			if (continuing_conditions_passing) {
+				fuseFakePosition();
+
+				const bool is_fusion_failing = isTimedOut(_time_last_hor_pos_fuse, (uint64_t)4e5);
+
+				if (is_fusion_failing) {
+					resetFakePosFusion();
+				}
+
+			} else {
+				stopFakePosFusion();
+			}
+
+		} else {
+			if (starting_conditions_passing) {
+				startFakePosFusion();
+
+				if (_control_status.flags.tilt_align) {
+					// The fake position fusion is not started for initial alignement
+					_warning_events.flags.stopping_navigation = true;
+					ECL_WARN("stopping navigation");
+				}
+			}
+		}
+	}
+}
+
+void Ekf::startFakePosFusion()
+{
+	if (!_using_synthetic_position) {
+		_using_synthetic_position = true;
+		_fuse_hpos_as_odom = false; // TODO: needed?
+		resetFakePosFusion();
+	}
+}
+
+void Ekf::resetFakePosFusion()
+{
+	_last_known_posNE = _state.pos.xy();
+	resetHorizontalPosition();
+	resetVelocity();
+	_time_last_hor_pos_fuse = _time_last_imu;
+}
+
+void Ekf::stopFakePosFusion()
+{
+	_using_synthetic_position = false;
+}
+
+void Ekf::fuseFakePosition()
+{
+	Vector3f fake_pos_obs_var;
+
+	if (_control_status.flags.in_air && _control_status.flags.tilt_align) {
+		fake_pos_obs_var(0) = fake_pos_obs_var(1) = sq(fmaxf(_params.pos_noaid_noise, _params.gps_pos_noise));
```

Comment:
Why is gps_pos_noise considered here?

Reply:
Nevermind, I think I understand this in the context of GPS now (reading below).

Reply:
That's something I didn't touch (just wanted to refactor here), but I think it's just to avoid having a smaller variance when fusion a fake position than a real one. Not sure if it's really needed to check that at this level. We could make sure we don't misconfigure it in the EKF2 module directly.